src/analysis/ast_visitor.rs

```diff
@@ -0,0 +1,985 @@
+use crate::clarity::functions::define::DefineFunctions;
+use crate::clarity::functions::NativeFunctions;
+use crate::clarity::representations::SymbolicExpressionType::*;
+use crate::clarity::representations::{Span, TraitDefinition};
+use crate::clarity::types::{TraitIdentifier, Value};
+use crate::clarity::{ClarityName, SymbolicExpression};
+use std::collections::HashMap;
+
+pub struct TypedVar<'a> {
+    pub name: &'a ClarityName,
+    pub type_expr: &'a SymbolicExpression,
+    pub decl_span: Span,
+}
+
+pub trait ASTVisitor<'a> {
+    fn traverse_expr(&mut self, expr: &'a SymbolicExpression) -> bool {
+        match &expr.expr {
+            AtomValue(value) => self.visit_atom_value(expr, value),
+            Atom(name) => self.visit_atom(expr, name),
+            List(exprs) => self.traverse_list(expr, &exprs),
+            LiteralValue(value) => self.visit_literal_value(expr, value),
+            Field(field) => self.visit_field(expr, field),
+            TraitReference(name, trait_def) => self.visit_trait_reference(expr, name, trait_def),
+        }
+    }
+
+    // AST level traverse/visit methods
+
+    fn traverse_list(
+        &mut self,
+        expr: &'a SymbolicExpression,
+        list: &'a [SymbolicExpression],
+    ) -> bool {
+        let mut rv = true;
+        if let Some((function_name, args)) = list.split_first() {
+            if let Some(function_name) = function_name.match_atom() {
+                if let Some(define_function) = DefineFunctions::lookup_by_name(function_name) {
+                    rv = match define_function {
+                        // TODO: I think all of these can be simplified to just use unwrap, since the structure has already been checked at this point.
```

Comment:
Correct, if this pass is running after the type checker, then you have the guaranty that we are looking at a contract correctly formatted.

Reply:
That being said, the thing I like with this approach, is that it could be a vehicle for future work, helping us working around the constraint "1 error raised per analysis". In the context of the stacks blockchain, there's no reason to continue parsing the contract if it has at least one error. In the context of developer tooling, we don't have to follow the same rules. What do you think?

Reply:
That depends on how good of a job we can do on working around errors to maintain an AST that is close enough to meaningfully continue.

Reply:
We would probably abort any child nodes checks, however we could maybe continue with the siblings / parents nodes?

Reply:
But yes, I agree. We should report as many errors as possible in one pass.

Reply:
Thinking about this some more now. I think that the check for a properly formed AST has to happen before this AST visitor, and in the case that we want to continue processing past an error, we'll need to insert a placeholder that maintains a well-formed AST, otherwise, it is too complex to handle in all later passes. That means that the unwrap technique should be fine here.

Reply:
Addressed in ed03067

Reply:
Force-pushed over that commit, it is now in 44e4abb